framer/Color.coffee

```diff
@@ -1,3 +1,468 @@
-class Color extends BaseClass
+{BaseClass} = require "./BaseClass"
 
-	constructor: ->
\ No newline at end of file
+class exports.Color extends BaseClass
+	constructor: (@color) ->
+
+		color = @color
+
+		# If input already is a Color object return input
+		if (color instanceof Color) then return color
+
+		# Convert input to RGB
+		rgb = inputToRGB(color)
+
+		@_r = rgb.r
+		@_g = rgb.g
+		@_b = rgb.b
+		@_a = rgb.a
+		@_roundA = Math.round(100*@_a) / 100
+
+	@define "r",
+		get: -> @_r
+		set: (value) -> throw "r is readonly"
+
+	@define "g",
+		get: -> @_g
+		set: (value) -> throw "g is readonly"
+
+	@define "b",
+		get: -> @_b
+		set: (value) -> throw "b is readonly"
+
+	@define "a",
+		get: -> @_a
+		set: (value) -> throw "a is readonly"
+
+	toHex: (allow3Char) ->
+		return rgbToHex(@_r, @_g, @_b, allow3Char)
+
+	toHexString: (allow3Char) ->
+		return '#' + @toHex(allow3Char)
+
+	toRgb: ->
+		r: Math.round(@_r)
+		g: Math.round(@_g)
+		b: Math.round(@_b)
+		a: @_a
+
+	toRgbString: ->
+		if @_a == 1 then "rgb(#{Utils.round(@_r, 0)}, #{Utils.round(@_g, 0)}, #{Utils.round(@_b, 0)})"
+		else "rgba(#{Utils.round(@_r, 0)}, #{Utils.round(@_g, 0)}, #{Utils.round(@_b, 0)}, #{@_roundA})"
+
+	toHsl: ->
+		hsl = rgbToHsl(@_r, @_g, @_b)
+		return { h: hsl.h * 360, s: hsl.s, l: hsl.l, a: @_a }
+
+	toHslString: ->
+		hsl = rgbToHsl(@_r, @_g, @_b)
+		h = Math.round(hsl.h * 360)
+		s = Math.round(hsl.s * 100)
+		l = Math.round(hsl.l * 100)
+		if @_a == 1 then "hsl(#{h}, #{s}%, #{l}%)"
+		else "hsl(#{h}, #{s}%, #{l}%, #{@_roundA})"
+
+	toName: ->
+		if @_a is 0 then return "transparent"
+		if @_a < 1 then return false
+		return cssNames[rgbToHex(@_r, @_g, @_b, @)] or false
+
+	lighten: (amount) ->
+		hsl = @toHsl()
+		hsl.l += amount / 100
+		hsl.l = Math.min(1, Math.max(0, hsl.l))
+		return new Color(hsl)
+
+	brighten: (amount) ->
+		rgb = @toRgb()
+		rgb.r = Math.max(0, Math.min(255, rgb.r - Math.round(255 * -(amount / 100))))
+		rgb.g = Math.max(0, Math.min(255, rgb.g - Math.round(255 * -(amount / 100))))
+		rgb.b = Math.max(0, Math.min(255, rgb.b - Math.round(255 * -(amount / 100))))
+		return new Color(rgb)
+
+	darken: (amount) ->
+		hsl = @toHsl()
+		hsl.l -= amount / 100
+		hsl.l = Math.min(1, Math.max(0, hsl.l))
+		return new Color(hsl)
+
+	desaturate: (amount) ->
+		hsl = @toHsl()
+		hsl.s -= amount / 100
+		hsl.s = Math.min(1, Math.max(0, hsl.s))
+		new Color(hsl)
+
+	saturate: (amount) ->
+		hsl = @toHsl()
+		hsl.s += amount / 100
+		hsl.s = Math.min(1, Math.max(0, hsl.s))
+		return new Color(hsl)
+
+	greyscale: ->
+		hsl = @toHsl()
+		return new Color(hsl).desaturate(100)
+
+	toString: ->
+		return @toRgbString()
+
+	copy: ->
+		copy = new Color
+			r: @_r
+			g: @_g
+			b: @_b
+			a: @_a
+		return copy
+
+	blend: (colorB, fraction) ->
+		return Color.blend(@, colorB, fraction)
+
+	##############################################################
+	## Class methods
+
+	@blend: (colorA, colorB, fraction) ->
```

Comment:
It would be nice to have a Color.random(opacity) method. Make Utils.randomColor(opacity) use that one.